### PR TITLE
control: improving and optimising the device finder

### PR DESCRIPTION
**Pull request by Djelibeybi**
This is the first WIP commit for some changes to device finder
that allow it to return the label for a switch. As requested initially,
this is now done by detecting the type of device first, then
sending either GetColor (for lights) or GetLabel (for switches).

I have adjusted most of the tests accordingly, except for
device_finder/test_information_retrieval.py. I am still trying to work
out how these work.

Signed-off-by: Avi Miller <me@dje.li>

### CODE REVIEW

**Comment by Djelibeybi** [comment]
GitHub imported the commit comments to the PR. Nifty.

**Comment by delfick** [comment]
> GitHub imported the commit comments to the PR. Nifty.

that is nifty :)

**Comment by delfick** [comment]
the comments aren't inline though which is less nifty hahhaha

**Comment by delfick** [comment]
also, goddamnit github markdown for removing the first line if a triple backtick has any text after it on the same line as the triple backtick

**Comment by Djelibeybi** [comment]
> the comments aren't inline though which is less nifty hahhaha

Yep and the commit view of the PR doesn't show them either. My bad. Let me work on the comments you've already made and then we can continue via the PR.

**Comment by Djelibeybi** [comment]
I'll fix the interactor tests shortly.

**Comment by Djelibeybi** [comment]
I haven't squashed the commits in case we want to cherry-pick a subset of initial commits.

**Comment by delfick** [comment]
cool cool. I'll find time to has look maybe tonight or tomorrow :)

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -7,6 +7,7 @@ on: [push, pull_request]
 jobs:
   tests:
     runs-on: ${{ matrix.os }}
+    if: github.event.pull_request.draft == false
```
Comment (delfick) [inline .github/workflows/ci.yml:10]: I have mixed feelings about this. Let's put it in a different PR and discuss there

File: apps/interactor/tests/commander/commands/test_control.py [hunk 2]
```diff
@@ -224,7 +224,13 @@ async def server(
         )
         for device in devices:
             if device is not kitchen_light:
-                devices.store(device).assertNoSetMessages()
+                assert devices.store(device).count(
```
Comment (delfick) [inline apps/interactor/tests/commander/commands/test_control.py:227]: why?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (Djelibeybi) [reply .github/workflows/ci.yml:10]: It's wrong: I meant `!=` false, i.e. it should only run tests on non-draft PRs.

File: apps/interactor/tests/conftest.py [hunk 3]
```diff
@@ -332,6 +336,7 @@ async def finish(self, exc_typ=None, exc=None, tb=None):
         color=hp.Color(100, 0, 1, 2500),
         group={"label": group_three_label, "identity": group_three_uuid},
         location={"label": location_two_label, "identity": location_two_uuid},
+        product=Products[VendorRegistry.LIFX, 22],
```
Comment (delfick) [inline apps/interactor/tests/conftest.py:339]: this shouldn't be necessary

File: docs/start/interacting/device_finder.rst [hunk 4]
```diff
@@ -139,7 +139,7 @@ in when you create the ``DeviceFinder``:
             finder = Finder(sender)
 
             # Turn off the lights with label kitchen
-            reference = DeviceFinder.from_kwargs(label="kitcehn", finder=finder)
+            reference = DeviceFinder.from_kwargs(label="kitchen", finder=finder)
```
Comment (delfick) [inline docs/start/interacting/device_finder.rst:142]: lol

File: .github/workflows/ci.yml (hunk 1 above)
Comment (Djelibeybi) [reply .github/workflows/ci.yml:10]: Sure.

File: modules/photons_control/device_finder.py [hunk 5]
```diff
@@ -366,10 +367,22 @@ def matches_all(self):
...
+            if e.value.condition is not None and not e.value.condition(self):
+                continue
+
             for key in e.value.keys:
                 if self[key] != sb.NotSpecified:
                     yield e
 
+    @property
+    def product_type(self):
+        """Set the product type based on which fields are being filtered"""
```
Comment (delfick) [inline modules/photons_control/device_finder.py:379]: we should base this off the cap on the device rather than what fields have values because the cap is the source of truth and the fields are more a heuristic in the context of a bug making them have values when they shouldn't

File: modules/photons_control/device_finder.py [hunk 6]
```diff
@@ -449,28 +462,43 @@ def from_options(kls, options, finder=None):
...
 
 
 class InfoPoints(enum.Enum):
     """
-    Enum used to determine what information is required for what keys
+    Enum used to determine which packet should be sent for each device type
+    to return the right data for the specified keys
     """
 
+    VERSION = Point(DeviceMessages.GetVersion(), ["product_id", "cap"], None)
```
Comment (delfick) [inline modules/photons_control/device_finder.py:478]:
```suggestion
    VERSION = Point(DeviceMessages.GetVersion(), ["product_id", "vendor_id"], None)
```

surely

Comment (delfick) [reply modules/photons_control/device_finder.py:478]: hmm, I thought I was adding reviews rather than making the comments, ignore this one

File: apps/interactor/tests/commander/commands/test_control.py (hunk 2 above)
Comment (Djelibeybi) [reply apps/interactor/tests/commander/commands/test_control.py:227]: My best guess is to do with the caching of GetVersion and the fact that all the devices are Lights, so the only packet type that is resent regularly is GetColor and at this point in the testing, the values are stale.

File: modules/photons_control/device_finder.py [hunk 7]
```diff
@@ -541,13 +576,86 @@ def location_id(self):
...
+        if self.product_id is sb.NotSpecified:
+            return sb.NotSpecified
+        return self.product.friendly
+
+    @property
+    def product_type(self):
+        if self.product is not sb.NotSpecified:
+            return (
+                ProductType.LIGHT.value
+                if Products[VendorRegistry.LIFX, self.product.pid].cap.is_light
```
Comment (delfick) [inline modules/photons_control/device_finder.py:596]:
```suggestion
                if self.cap.is_light
```

Comment (delfick) [reply modules/photons_control/device_finder.py:596]: can even remove the check on self.product

Comment (delfick) [reply modules/photons_control/device_finder.py:596]: and check if cap is unknown for returning ProductType.UNKNOWN rather than unspecified

File: apps/interactor/tests/conftest.py (hunk 3 above)
Comment (Djelibeybi) [reply apps/interactor/tests/conftest.py:339]: Confirmed. Will remove the line from each block.

File: modules/photons_control/device_finder.py [hunk 8]
```diff
@@ -541,13 +576,86 @@ def location_id(self):
...
+                else ProductType.SWITCH.value
+            )
+        else:
+            return sb.NotSpecified
+
+    @property
+    def firmware_version(self):
+        if self.firmware is sb.NotSpecified:
+            return sb.NotSpecified
+        return str(self.firmware)
```
Comment (delfick) [inline modules/photons_control/device_finder.py:606]: I'd say remove your `__str__` method and return `tuple(self.firmware)` and add an

```
def __iter__(self):
   yield from (self.major, self.minor)
```
Screen Shot 2022-01-02 at 9 18 20 pm


To the firmware object

because firmware version is a tuple of two numbers and should be treated as such.

File: modules/photons_control/device_finder.py (hunk 5 above)
Comment (Djelibeybi) [reply modules/photons_control/device_finder.py:379]: This is for the Filter() class, and I figured folks would want to filter on device settings, rather than capabilties.

File: modules/photons_control/device_finder.py [hunk 9]
```diff
@@ -541,13 +576,86 @@ def location_id(self):
...
+
+    @property
+    def cap(self):
```
Comment (delfick) [inline modules/photons_control/device_finder.py:609]: let's change, in a PR before this one this property to be like `abilities` so that we can replace it with a cap that is `self.product.cap(*self.firmware_version)` so it's an actual cap object

File: modules/photons_control/device_finder.py [hunk 10]
```diff
@@ -541,13 +576,86 @@ def location_id(self):
...
+            return Products[VendorRegistry.LIFX, self.product.pid].cap
+        else:
+            return sb.NotSpecified
+
     def as_dict(self):
         actual = super(Device, self).as_dict()
         del actual["group"]
         del actual["limit"]
         del actual["location"]
+        del actual["product"]
```
Comment (delfick) [inline modules/photons_control/device_finder.py:654]: does actual have product and firmware at this point?

File: modules/photons_control/device_finder.py [hunk 11]
```diff
@@ -704,11 +802,26 @@ async def matches(self, sender, fltr, collections):
 
         async def gen(reference, sender, **kwargs):
             for e in self.points_from_fltr(fltr):
+                if e == InfoPoints.VERSION:
```
Comment (delfick) [inline modules/photons_control/device_finder.py:805]: ?

File: modules/photons_control/device_finder.py [hunk 12]
```diff
@@ -704,11 +802,26 @@ async def matches(self, sender, fltr, collections):
...
+                    continue
+                if e.value.condition is not None and not e.value.condition(self):
+                    continue
                 if self.final_future.done():
                     return
                 if not self.point_futures[e].done():
                     yield e.value.msg
 
+        for e in self.points_from_fltr(fltr):
```
Comment (delfick) [inline modules/photons_control/device_finder.py:814]: ?

File: modules/photons_control/device_finder.py (hunk 7 above)
Comment (Djelibeybi) [reply modules/photons_control/device_finder.py:596]: So I did that initially, but it required waiting for enough response which meant putting logic in weird places. So I moved the logic to determine the best current value based off what we know from the packets we have thus far.

File: modules/tests/photons_control_tests/device_finder/test_device.py [hunk 13]
```diff
@@ -57,12 +86,31 @@ def device(self):
             "kelvin": 2500,
             "firmware_version": "1.2",
             "product_id": 22,
-            "cap": ["multizone", "color"],
+            "product_name": "LIFX Color 1000",
+            "product_type": "light",
+            "cap": [
```
Comment (delfick) [inline modules/tests/photons_control_tests/device_finder/test_device.py:91]: https://github.com/delfick/photons/blob/main/modules/photons_pytest.py#L504

File: modules/tests/photons_control_tests/device_finder/test_device.py [hunk 14]
```diff
@@ -75,22 +123,21 @@ def assertChange(field, value):
             assert device.info == info
             assert device.as_dict() == expected
 
+        assertChange("product", Products[VendorRegistry.LIFX, 22])
+        assertChange("firmware", hp.Firmware(1, 2, 0, 0))
```
Comment (delfick) [inline modules/tests/photons_control_tests/device_finder/test_device.py:127]:
```suggestion
        assertChange("firmware", hp.Firmware(1, 2))
```

File: modules/tests/photons_control_tests/device_finder/test_filter.py [hunk 15]
```diff
@@ -261,7 +261,7 @@ def assertFltrMatches(self, filtr, expect):
     describe "points":
         it "returns the InfoPoint enums for the fields that have values":
             filtr = Filter.from_kwargs(label="kitchen", product_id=22)
-            assert set(filtr.points) == set([InfoPoints.LIGHT_STATE, InfoPoints.VERSION])
+            assert set(filtr.points) == set([InfoPoints.LABEL, InfoPoints.VERSION])
```
Comment (delfick) [inline modules/tests/photons_control_tests/device_finder/test_filter.py:264]: shouldn't a 22 get LIGHT_STATE, not LABEL?

File: modules/photons_control/device_finder.py (hunk 10 above)
Comment (Djelibeybi) [reply modules/photons_control/device_finder.py:654]: Sometimes, yes.

File: modules/photons_control/device_finder.py (hunk 11 above)
Comment (Djelibeybi) [reply modules/photons_control/device_finder.py:805]: This is (I'd guess) attempt 1 billiionth and 3 at getting the futures to work properly such that the information retrieval test didn't run into oblivion sending the same packets over and over. There is possibly a more elegant way to do it without breaking the InfoPoints enum, but I couldn't find it at the time.

File: modules/photons_control/device_finder.py (hunk 12 above)
Comment (Djelibeybi) [reply modules/photons_control/device_finder.py:814]: Attempt one billionth and 2.

File: modules/photons_control/device_finder.py (hunk 7 above)
Comment (delfick) [reply modules/photons_control/device_finder.py:596]: but self.product should give back either the product or an instance of https://github.com/delfick/photons/blob/main/modules/photons_products/base.py#L228 (which we should make better, in a pre-PR) and self.cap should be based off self.product.

File: modules/photons_control/device_finder.py (hunk 5 above)
Comment (delfick) [reply modules/photons_control/device_finder.py:379]: the point of the product system is that it's accurate and all encompassing so this should be based off that rather than how it's implemented

File: modules/tests/photons_control_tests/device_finder/test_filter.py (hunk 15 above)
Comment (Djelibeybi) [reply modules/tests/photons_control_tests/device_finder/test_filter.py:264]: No, because it doesn't need to know if it's a light or not. It's filtering on label and product ID and that doesn't need a GetColor/is_light.

File: modules/tests/photons_control_tests/device_finder/test_device.py (hunk 13 above)
Comment (Djelibeybi) [reply modules/tests/photons_control_tests/device_finder/test_device.py:91]: I think I discoved this after manually changing all the test devices and then forgot to go back to see if it would solve my previous problems.

File: modules/photons_control/device_finder.py (hunk 12 above)
Comment (delfick) [reply modules/photons_control/device_finder.py:814]: perhaps once we've sorted out the other comments I'll take a look at what's happening here. Can't remember how device_finder works off the top of my head well enough to remember what this does and why it does it.

File: modules/tests/photons_control_tests/device_finder/test_filter.py (hunk 15 above)
Comment (delfick) [reply modules/tests/photons_control_tests/device_finder/test_filter.py:264]: right. nice.

**Comment by Djelibeybi** [comment]
I did some performance testing: https://github.com/delfick/photons/discussions/85

File: modules/photons_control/device_finder.py (hunk 9 above)
Comment (delfick) [reply modules/photons_control/device_finder.py:607]: call it abilities rather than cap so you can make a property called cap that is a cap object rather than a list of strings.

Comment (Djelibeybi) [reply modules/photons_control/device_finder.py:607]: Ah, gotcha. I thought there was another property called `abilities` somewhere in the desired format. :) The property `capabilties` is already a `cap` object. I only kept `cap` for backwards compatibility.

Comment (delfick) [reply modules/photons_control/device_finder.py:607]: the purpose of the list of strings remains but the word cap is "soft-reserved" as such for the cap object and it was a mistake for me to name this method cap. Hence a separate PR before this one to rename this property to abilities so that cap is not used for this PR and this PR can then create a cap property that returns a cap object

Comment (Djelibeybi) [reply modules/photons_control/device_finder.py:607]: Righto.

**Comment by delfick** [comment]
resolved in other PRs. Thanks for doing the prototyping!
